src/backends/climacore/tendencies.jl

```diff
@@ -122,45 +56,57 @@ function create_rhs(::ClimaCoreBackend, model::HydrostaticModel, function_space:
         dw = dYf.w
 
         # auxiliary calculations
-        u_1 = parent(u)[1]
-        v_1 = parent(v)[1]
-        u_wind = sqrt(u_1^2 + v_1^2)
+        center_space = axes(Yc)
+        dz_top = center_space.face_local_geometry.WJ[end]
+        u_top, v_top = parent(u)[end], parent(v)[end]
+        u_btm, v_btm = parent(u)[1], parent(v)[1]
+        u_wind = sqrt(u_btm^2 + v_btm^2)
 
         # density (centers)
-        gradc2f = Operators.GradientC2F()
-        gradf2c = Operators.GradientF2C(bottom = Operators.SetValue(0.0), top = Operators.SetValue(0.0))
+        flux_btm = 0.0
+        flux_top = 0.0
+        ∇c = Operators.GradientC2F()
+        ∇f = Operators.GradientF2C(bottom = Operators.SetValue(flux_btm), top = Operators.SetValue(flux_top))
 
         If = Operators.InterpolateC2F(bottom = Operators.Extrapolate(), top = Operators.Extrapolate())
-        @. dρ = gradf2c( -w * If(ρ) ) # Eq. 4.11
+        @. dρ = ∇f( -w * If(ρ) ) # Eq. 4.11
 
         # potential temperature (centers)
-        gradc2f = Operators.GradientC2F()
-        gradf2c = Operators.GradientF2C(bottom = Operators.SetValue(0.0), top = Operators.SetValue(0.0)) # Eq. 4.20, 4.21
+        flux_btm = 0.0
+        flux_top = 0.0
+        ∇c = Operators.GradientC2F()
+        ∇f = Operators.GradientF2C(bottom = Operators.SetValue(flux_btm), top = Operators.SetValue(flux_top)) # Eq. 4.20, 4.21
 
-        @. dρθ = gradf2c( -w * If(ρθ) + ν * gradc2f(ρθ/ρ) ) # Eq. 4.12
+        @. dρθ = ∇f( -w * If(ρθ) + ν * ∇c(ρθ/ρ) ) # Eq. 4.12
 
-        # u velocity (centers)
-        gradc2f = Operators.GradientC2F(top = Operators.SetValue(ug)) # Eq. 4.18
-        gradf2c = Operators.GradientF2C(bottom = Operators.SetValue(Cd * u_wind * u_1)) # Eq. 4.16
-        
-        A = Operators.AdvectionC2C(bottom = Operators.SetValue(0.0), top = Operators.SetValue(0.0))
-        @. du = gradf2c(ν * gradc2f(u)) + f * (v - vg) - A(w, u) # Eq. 4.8
+        # u velocity (center
+        flux_btm = Cd * u_wind * u_btm
+        flux_top = ν * (v_top - vg) / dz_top / 2
+        ∇c = Operators.GradientC2F() # Eq. 4.18
+        ∇f = Operators.GradientF2C(bottom = Operators.SetValue(flux_btm), top = Operators.SetValue(flux_top)) # Eq. 4.16
```

Comment:
It looks like you are setting state BC on v (v_top face = vg), and on u (u_top face = ug), and then a flux BC at the bottom for the u,v equations.  (and the A operator does not add any extra constraints?)

so would the alternative be to set:
```gradc2f = Operators.GradientC2F(top  = setvalue(ug))
gradf2c = Operators.GradientF2C(bottom = setvalue(flux_btm)
@. du = gradf2c(nu*gradc2f(u))```?

and here you are explicitly making the top state BC into a flux BC as we talked about?

I dont quite understand the chain here - seeing your idea for the interface will help. Is it: (1) the user sets a state BC, (2) internally there is an intermediate function that maps that into a flux bc using something appropriate/unique for the problem and  then (3) that flux bc is passed to the create RHS function which *only* handles flux bcs?

Reply:
is there any circumstance where approximating the derivative like this (to get a flux bc from state bc) would be inconsistent with what you would get if you explicitly set the BC like I wrote above?

Reply:
Yes, all that you wrote is what I had in mind here. The use picks some boundary condition step (1), then that boundary condition corresponds to a certain flux boundary conditions internally (2), and that flux is invoked here (3).